src/main/java/net/neoforged/neoforge/attachment/AttachmentType.java

```diff
@@ -105,118 +99,92 @@ public static <T> Builder<T> builder(Supplier<T> defaultValueSupplier) {
+        return Builder.fromReflection(holderClass, defaultValueConstructor);
     }
 
-    /**
-     * Create a builder for an attachment type that uses {@link INBTSerializable} for serialization.
-     * Other kinds of serialization can be implemented using {@link #builder(Supplier)} and {@link Builder#serialize(IAttachmentSerializer)}.
-     *
-     * <p>This overload allows capturing a reference to the {@link IAttachmentHolder} for the attachment.
-     * To obtain a specific subtype, the holder can be cast.
-     * If the holder is of the wrong type, the constructor should throw an exception.
-     * See {@link #serializable(Supplier)} for an overload that does not capture the holder.
-     */
-    public static <S extends Tag, T extends INBTSerializable<S>> Builder<T> serializable(Function<IAttachmentHolder, T> defaultValueConstructor) {
-        return builder(defaultValueConstructor).serialize(new IAttachmentSerializer<S, T>() {
-            @Override
-            public T read(IAttachmentHolder holder, S tag, HolderLookup.Provider provider) {
-                var ret = defaultValueConstructor.apply(holder);
-                ret.deserializeNBT(provider, tag);
-                return ret;
-            }
-
-            @Nullable
-            @Override
-            public S write(T attachment, HolderLookup.Provider provider) {
-                return attachment.serializeNBT(provider);
-            }
-        });
-    }
-
-    public static class Builder<T> {
-        private final Function<IAttachmentHolder, T> defaultValueSupplier;
+    public static class Builder<T, P extends IAttachmentHolder> {
+        private final Function<P, T> defaultValueSupplier;
+        private final Class<P> attachmentHolderClass;
+        private final Class<T> dataType;
+        @Nullable
+        private Encoder<T> serializer;
         @Nullable
-        private IAttachmentSerializer<?, T> serializer;
+        private Decoder<T> deserializer;
+        BiConsumer<P, T> postDeserialize;
         private boolean copyOnDeath;
+        private Predicate<T> shouldSerialize;
         @Nullable
-        private IAttachmentCopyHandler<T> copyHandler;
+        private StreamCodec<RegistryFriendlyByteBuf, T> copyHandler;
 
-        private Builder(Function<IAttachmentHolder, T> defaultValueSupplier) {
+        private Builder(Class<P> parentClass, Class<T> dataType, Function<P, T> defaultValueSupplier) {
+            this.attachmentHolderClass = parentClass;
+            this.dataType = dataType;
             this.defaultValueSupplier = defaultValueSupplier;
+            this.shouldSerialize = Predicates.alwaysTrue();
+            this.postDeserialize = (holder, inst) -> {};
+        }
+
+        private static <T, P extends IAttachmentHolder> Builder<T, P> fromReflection(Function<P, T> defaultValueSupplier) {
+            Class<?>[] typeArgs = TypeResolver.resolveRawArguments(Function.class, defaultValueSupplier.getClass());
```

Comment:
Ugh yeah no, this is even worse than the other one. Way to break in _all_ sorts of cases when people have generics such that this isn't a recoverable type arg... let's _please_ not do this.